From 46c43ce7fc3126d31d0fe69fe4931d98bcfc246c Mon Sep 17 00:00:00 2001
From: "mujiburrahimank@maxlinear.com"
 <mujiburrahimank@lvdocker7.lvsw.maxlinear.com>
Date: Sun, 17 Aug 2025 23:47:00 -0700
Subject: [PATCH] WLANRTSYS-93122 : ACS - 2G Interface goes down as CSA is sent
 for same channel

Issue : 2G Interface goes down after 60 minutes (once bgscan is completed)

RCA   : After ACS is completed due to bgscan interval the results in ACS
smart info file for 2G is worse as the utilisation factor is below
threshold for all candidates due to which it sends CSA to the current
configured channel.

Fix   : When utilisation factor is below threshold for all candidates
consider channel switching where switch to current channel is not
allowed.

Fixup : WLANRTSYS-50170 [MXL] Smart ACS implementation
---
 src/ap/mxl_acs_proprietary.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/src/ap/mxl_acs_proprietary.c b/src/ap/mxl_acs_proprietary.c
index 842df3dc5..fe9cadbb0 100644
--- a/src/ap/mxl_acs_proprietary.c
+++ b/src/ap/mxl_acs_proprietary.c
@@ -2711,7 +2711,8 @@ static bool mxl_acs_check_new_chan_is_ok(struct hostapd_iface *iface, int min_ra
 					selected_cand_rank);
 		}
 		if (candidates[iface->mxl_iface.selected_candidate].width != actual_width ||
-		   (candidates[iface->mxl_iface.selected_candidate].exclude_reason != ACS_NO_EXCLUSION) ||
+		   ((candidates[iface->mxl_iface.selected_candidate].exclude_reason != ACS_NO_EXCLUSION) &&
+			(candidates[iface->mxl_iface.selected_candidate].exclude_reason != ACS_EXCLUDE_LOW_PRIMARY_UTIL)) ||
 		   (candidates[iface->mxl_iface.selected_candidate].rank == -1)) {
 			wpa_printf(MSG_WARNING, "invalid previously candidate %d detected, width %d",
 					iface->mxl_iface.selected_candidate, actual_width);
@@ -3689,8 +3690,10 @@ int mxl_acs_recalc_ranks_and_set_chan (struct hostapd_iface *iface, int switch_r
 			candidates[i].bw_comp_rank = candidates[i].rank - tx_power_rank - num_bss_rank;
 			acs_calculate_rank_utilization(&candidates[i], rank_util_primary);
 
-			if (candidates[i].rank_util_primary <= ACS_RANK_UTIL_MIN)
-				candidates[i].exclude_reason = ACS_EXCLUDE_LOW_PRIMARY_UTIL;
+			if (candidates[i].rank_util_primary <= ACS_RANK_UTIL_MIN) {
+				if (candidates[i].exclude_reason != ACS_EXCLUDE_DISABLED)
+					candidates[i].exclude_reason = ACS_EXCLUDE_LOW_PRIMARY_UTIL;
+			}
 		} else {
 			candidates[i].rank = -1;
 			candidates[i].rank_util_primary = -1;
-- 
2.43.0

